### ISSUE TITLE
Cosmos DB account virtual network filter rule doesn't work if deployed from Terraform

### ISSUE THREAD

**Issue by jjoynsonerwin** [hashicorp/terraform-provider-azurerm#9977]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform 0.13.5
AzureRM 2.37.0

### Affected Resource(s)

* `azurerm_cosmosdb_account`

### Terraform Configuration Files

```hcl
resource "azurerm_cosmosdb_account" "example" {
  name                = etc
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name
  offer_type          = "Standard"
  kind                = "MongoDB"

  consistency_policy {
    consistency_level = "Strong"
  }

  geo_location {
    location          = azurerm_resource_group.example.location
    failover_priority = 0
  }

  enable_automatic_failover     = true
  public_network_access_enabled = false

  capabilities {
    name = "EnableAggregationPipeline"
  }

  capabilities {
    name = "mongoEnableDocLevelTTL"
  }

  capabilities {
    name = "EnableMongo"
  }

  is_virtual_network_filter_enabled = true

  virtual_network_rule {
    id = <MY SUBNET ID>
  }

  tags = local.common_tags
}
```

### Debug Output

N/A

### Panic Output

N/A

### Expected Behaviour

Virtual Network rule should be added to the CosmosDB account and should allow traffic from the Virtual network. 

### Actual Behaviour

Virtual Network rule is added to the CosmosDB account as expected (viewed from the portal).
However traffic from the Virtual network is blocked. Get an error saying "Request originated from client VNET through service endpoint. This is blocked by your Cosmos DB account firewall settings."

If I removed the vnet filtering lines:

```hcl
  is_virtual_network_filter_enabled = true

  virtual_network_rule {
    id = <MY SUBNET ID>
  }
```
then deploy the infrastructure and add the same vnet rule from the portal then the traffic is allowed and I do not get an error.

TLDR: Identical vnet filtering settings (As viewed from the portal) behave differently if deployed from TF vs from the portal.

### Steps to Reproduce

Run the Terraform code above with a virtual network, subnet and private endpoints. 

```hcl
resource "azurerm_virtual_network" "example" {
  name                = etc
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location
  address_space       = ["10.0.0.0/8"]
  tags                = local.common_tags
}

resource "azurerm_subnet" "aks" {
  name                 = "snet-aks"
  resource_group_name  = azurerm_resource_group.example.name
  virtual_network_name = azurerm_virtual_network.example.name
  address_prefixes = [
    "10.1.0.0/16"
  ]
  service_endpoints = [
    "Microsoft.AzureCosmosDB"
  ]

  enforce_private_link_endpoint_network_policies = true
}

resource "azurerm_private_endpoint" "cosmos" {
  name                = "pe-cosmos"
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location
  subnet_id           = azurerm_subnet.aks.id

  private_service_connection {
    name                           = "cosmos_private_endpoint"
    is_manual_connection           = "false"
    private_connection_resource_id = azurerm_cosmosdb_account.example.id
    subresource_names              = ["MongoDB"]
  }
}

resource "azurerm_private_dns_zone" "cosmos" {
  name                = "privatelink.mongo.cosmos.azure.com"
  resource_group_name = data.azurerm_resource_group.example.name
}

resource "azurerm_private_dns_a_record" "cosmos" {
  name                = azurerm_cosmosdb_account.example.name
  resource_group_name = data.azurerm_resource_group.example.name
  zone_name           = azurerm_private_dns_zone.example.name
  ttl                 = 10
  records             = [azurerm_private_endpoint.cosmos.private_service_connection.0.private_ip_address]
}

resource "azurerm_private_dns_zone_virtual_network_link" "cosmos" {
  name                  = azurerm_cosmosdb_account.example.name
  private_dns_zone_name = azurerm_private_dns_zone.cosmos.name
  resource_group_name   = data.azurerm_resource_group.example.name
  virtual_network_id    = data.azurerm_virtual_network.example.id
}


```

### Important Factoids

Running in uksouth

### References

Possibly related? https://github.com/MicrosoftDocs/azure-docs/issues/28461

**Comment by MattMencel** [comment]
@jjoynsonerwin did you ever come up with a solution? I'm running in to the same issue. In my case I am whitelisting an AKS subnet in the Cosmos firewall rules. But as soon as the firewall is enabled, access from AKS stops working.

**Comment by jjoynsonerwin** [comment]
@MattMencel not yet I'm afraid. 
What I have noticed is that it appears that the Azure portal is setting the Cosmos account public even tho it says it is private and limited to the vnet. 

Process:
1) Deploy as above.
2) Remove the vnet from the firewall rules.
3) Readd the same vnet via the portal > Database access via vnet now works.
4) Redeploy from Terraform and it shows changes to the cosmosdb_account as follows:

```
  ~ resource "azurerm_cosmosdb_account" "example" { 
      id = abc123
      name = example

      ~ public_network_access_enabled     = true -> false
```

I am planning to raise this with MS support when I get time.

**Comment by MattMencel** [comment]
I may open a support request as well. This is what I've noticed myself.

1. Deploy similar to yours, access is blocked from AKS.
2. Go to portal and "Allow access from Azure Portal". This is the `public_network_access_enabled` flag in Terraform. Now AKS can reach Cosmos.
3. Disable "Allow access from Azure Portal". AKS access still works so I thought that was the trick. However next time I run Terraform it tells me the `public_network_access_enabled` flag is still set to "true"... so something is not changing even though the portal has that box un-ticked now.

**Comment by ElvenSpellmaker** [comment]
`public_network_access_enabled` is not the same thing as `Allow access from Azure Portal`. In the ARM template if you enable `Allow access from Azure Portal` it adds a load of IPs to the template and that's it. `"publicNetworkAccess"` is the ARM equivalent `public_network_access_enabled` parameter.

**Comment by markbangert** [comment]
We have a similar problem. I think we were able to fix this issue using the following terraform configuration in our cosmosdb definition:

``` 
  public_network_access_enabled = true
  is_virtual_network_filter_enabled = true

  virtual_network_rule {
    id = azurerm_subnet.aks_subnet.id
  }
```

With this setup, we were able to confirm that the cosmos db blocks access from the general internet but it is reachable from our vnet. But why do I have to set `public_network_access_enabled = true`? This is not intuitive! If I set `public_network_access_enabled = false` access from my subnet does not work.

**Comment by fabian-ro** [comment]
We had a similar issue. If the Cosmos DB account is created with `public_network_access_enabled = false` then the firewall settings of the account shows "Allow access from: All networks" but public access does **not** work.
If the account is created with `public_network_access_enabled = true` the settings in the Azure portal look the same but public access actually works.

**Comment by singh-navpreet** [comment]
when we set   `public_network_access_enabled = false` it will restrict access from VNet and public network(even if IP's are defined in ip_range_filter).  
In this scenario only way to access cosmos database is by using private link.[https://docs.microsoft.com/en-us/azure/private-link/tutorial-private-endpoint-cosmosdb-portal](url).

So to make Vnet and IP filtering work , we need to set `public_network_access_enabled = true`. This does not means that our database will be accessible publicly and access will still be allowed only to IP's defined in ip_range_filter or via VNet integration.
Here are tf settings that worked for me:

`public_network_access_enabled = true
  is_virtual_network_filter_enabled = true
  virtual_network_rule {
    id = "/subscriptions/xxxxxxxxxxxxxx/resourceGroups/zzzzzzzz/providers/Microsoft.Network/virtualNetworks/xxxxxxxx/subnets/zzzzzzzz"
    ignore_missing_vnet_service_endpoint  = false
  }
  ip_range_filter = "some ip's"`

**Comment by jurkrit** [comment]
> when we set `public_network_access_enabled = false` it will restrict access from VNet and public network(even if IP's are defined in ip_range_filter). In this scenario only way to access cosmos database is by using private link.[https://docs.microsoft.com/en-us/azure/private-link/tutorial-private-endpoint-cosmosdb-portal](url). Here is what MS says when i tried to diagnose this issue before raising support request. ![Screenshot 2021-09-02 at 15 10 23](https://user-images.githubusercontent.com/50949810/131853230-525fb486-4fae-442a-9cae-9c69a428dc6c.png)
> 
> So to make Vnet and IP filtering work , we need to set `public_network_access_enabled = true`. This does not means that our database will be accessible publicly and access will still be allowed only to IP's defined in ip_range_filter or via VNet integration. But this does not make the firewall settings of the account in portal to show "Allow access from: All networks"!! . They showed Allow access from selected networks and my configured VNet and IP rules. Here are tf settings that worked for me:
> 
> `public_network_access_enabled = true` `is_virtual_network_filter_enabled = true` `virtual_network_rule { id = "/subscriptions/xxxxxxxxxxxxxx/resourceGroups/zzzzzzzz/providers/Microsoft.Network/virtualNetworks/xxxxxxxx/subnets/zzzzzzzz"` ` ignore_missing_vnet_service_endpoint = false }` `ip_range_filter = "some ip's"`

My config is the same as above but no vnets or firewall rules are added at all.

**Comment by noiano** [comment]
any updates? 🥲

**Comment by kinseii** [comment]
bump

**Comment by rcskosir** [comment]
Thanks for opening this issue. This was a problem in the 2.x version of the provider which is no longer actively maintained.
If this is still an issue with the 3.x version of the provider please do let us know by opening a new issue, thanks!

**Comment by ElvenSpellmaker** [comment]
> Thanks for opening this issue. This was a problem in the 2.x version of the provider which is no longer actively maintained. If this is still an issue with the 3.x version of the provider please do let us know by opening a new issue, thanks!

Shouldn't you confirm this is not an issue in 3.x before closing?
